File: app/src/main/java/com/android/mathias/velocity/ActivityCreateRoute.java
package com.android.mathias.velocity;

import android.Manifest;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import java.io.IOException;
import java.util.List;

public class ActivityCreateRoute extends AppCompatActivity implements
        OnMapReadyCallback,
        GoogleApiClient.ConnectionCallbacks,
        GoogleApiClient.OnConnectionFailedListener {

    private static final int PERMISSIONS_REQUEST_ACCESS_LOCATION = 100;
    protected static final String ROUTE_NAME = "ROUTE_NAME";
    protected static final String START_LOC = "START_LOC";
    protected static final String END_LOC = "END_LOC";
    protected static final String START_LOC_NAME = "START_LOC_NAME";
    protected static final String END_LOC_NAME = "END_LOC_NAME";
    protected static final String RESULT_BUNDLE = "RESULT_BUNDLE";

    private GoogleMap mMap;
    private GoogleApiClient mGoogleApiClient;
    private LatLng mStartLoc;
    private LatLng mEndLoc;
    private Geocoder mGeocoder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_route);
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        if(getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        mapFragment.getMapAsync(this);
        Button btnSave = (Button) findViewById(R.id.btn_save_route);
        btnSave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                promptForNameAndReturn();
            }
        });
        if (mGoogleApiClient == null) {
            mGoogleApiClient = new GoogleApiClient.Builder(this)
                    .addConnectionCallbacks(this)
                    .enableAutoManage(this, this)
                    .addOnConnectionFailedListener(this)
                    .addApi(LocationServices.API).build();
        }
        mGeocoder = new Geocoder(this.getApplicationContext());
    }

    private void promptForNameAndReturn() {
        final InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        new AlertDialog.Builder(this)
                .setTitle("Route name")
                .setView(R.layout.dialog_route)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        String routeName = ((EditText) ((Dialog) dialogInterface).findViewById(R.id.txt_name_prompt)).getText().toString();
                        finish(routeName);
                        imm.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0);
                    }
                }).show();
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
    }

    private void finish(String routeName) {
        if (mStartLoc != null && mEndLoc != null) {
            String startLocName = "Start";
            String endLocName = "End";
            try {
                startLocName = mGeocoder.getFromLocation(mStartLoc.latitude, mStartLoc.longitude, 1).get(0).getAddressLine(0);
                endLocName = mGeocoder.getFromLocation(mEndLoc.latitude, mEndLoc.longitude, 1).get(0).getAddressLine(0);
            } catch (IOException ex) {
                ex.printStackTrace();
            }
            Intent returnIntent = new Intent();
            Bundle bundle = new Bundle();
            bundle.putDoubleArray(START_LOC, new double[]{mStartLoc.latitude, mStartLoc.longitude});
            bundle.putDoubleArray(END_LOC, new double[]{mEndLoc.latitude, mEndLoc.longitude});
            bundle.putString(START_LOC_NAME, startLocName);
            bundle.putString(END_LOC_NAME, endLocName);
            bundle.putString(ROUTE_NAME, routeName);
            returnIntent.putExtra(String.valueOf(RESULT_BUNDLE), bundle);
            ActivityCreateRoute.this.setResult(RESULT_OK, returnIntent);
            ActivityCreateRoute.this.finish();
        } else {
            ActivityCreateRoute.this.setResult(RESULT_CANCELED);
            ActivityCreateRoute.this.finish();
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        mMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
            @Override
            public void onMapClick(LatLng latLng) {
                ActivityCreateRoute.this.handleMapClick(mMap, latLng);
            }
        });
        mMap.setOnMarkerDragListener(new GoogleMap.OnMarkerDragListener() {
            @Override
            public void onMarkerDragStart(Marker marker) { }
            @Override
            public void onMarkerDrag(Marker marker) { }
            @Override
            public void onMarkerDragEnd(Marker marker) {
                String address = null;
                try {
                    List<Address> addresses = mGeocoder.getFromLocation(marker.getPosition().latitude, marker.getPosition().longitude ,1);
                    address = addresses.get(0).getAddressLine(0);
                } catch (IOException e) { e.printStackTrace(); }
                if (address != null) {
                    marker.setPosition(marker.getPosition());
                    marker.setSnippet(address);
                    if (marker.getTitle().equals(getString(R.string.end_address_title))) {
                        mEndLoc = new LatLng(marker.getPosition().latitude, marker.getPosition().longitude);
                        ((TextView) findViewById(R.id.txt_end_address)).setText(String.format(getString(R.string.route_end_prefix), address));
                    } else {
                        mStartLoc = new LatLng(marker.getPosition().latitude, marker.getPosition().longitude);
                        ((TextView) findViewById(R.id.txt_start_address)).setText(String.format(getString(R.string.route_start_prefix), address));
                    }
                    marker.showInfoWindow();
                }
            }
        });
    }


    @Override
    public void onConnected(@Nullable Bundle bundle) {
        updateLocation();
    }

    private void handleMapClick(GoogleMap gMap, LatLng latLng) {
        try {
            List<Address> addresses = mGeocoder.getFromLocation(latLng.latitude, latLng.longitude ,1);
            if (addresses.size() == 0) { return; }
            String address = addresses.get(0).getAddressLine(0);
            if (mStartLoc == null) {
                Marker marker = gMap.addMarker(new MarkerOptions()
                        .position(latLng)
                        .title(getString(R.string.start_address_title))
                        .snippet(address)
                        .flat(true)
                        .draggable(true));
                mStartLoc = latLng;
                marker.showInfoWindow();
                TextView txtStartAddress = (TextView) findViewById(R.id.txt_start_address);
                txtStartAddress.setText(String.format(getString(R.string.route_start_prefix), address));
                txtStartAddress.setVisibility(View.VISIBLE);
            } else if (mEndLoc == null) {
                Marker marker = gMap.addMarker(new MarkerOptions()
                        .position(latLng)
                        .title(getString(R.string.end_address_title))
                        .snippet(address)
                        .flat(true)
                        .draggable(true));
                mEndLoc = latLng;
                marker.showInfoWindow();
                findViewById(R.id.btn_save_route).setVisibility(View.VISIBLE);
                TextView txtEndAddress = (TextView) findViewById(R.id.txt_end_address);
                txtEndAddress.setText(String.format(getString(R.string.route_end_prefix), address));
                txtEndAddress.setVisibility(View.VISIBLE);
            } else {
                mStartLoc = null;
                mEndLoc = null;
                gMap.clear();
                findViewById(R.id.btn_save_route).setVisibility(View.GONE);
                findViewById(R.id.txt_start_address).setVisibility(View.GONE);
                findViewById(R.id.txt_end_address).setVisibility(View.GONE);
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }


    }

    private void updateLocation() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, PERMISSIONS_REQUEST_ACCESS_LOCATION);
        } else {
            Location mLastKnownLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleApiClient);
            if (mLastKnownLocation != null) {
                LatLng lastKnown = new LatLng(mLastKnownLocation.getLatitude(), mLastKnownLocation.getLongitude());
                mMap.setMyLocationEnabled(true);
                mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(lastKnown, 14f));
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        switch (requestCode) {
            case PERMISSIONS_REQUEST_ACCESS_LOCATION: {
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    updateLocation();
                }
                // else {
                    // permission denied, boo! Disable the functionality that depends on this permission.
                // }
            }
        }
    }

    @Override
    public void onConnectionSuspended(int i) {
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/android/mathias/velocity/ActivityCreateRoute.kt
package com.android.mathias.velocity

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Geocoder
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.android.mathias.velocity.util.BottomSheetCreateRoutes
import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.google.android.gms.common.GooglePlayServicesRepairableException
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.places.ui.PlacePicker
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import java.io.IOException

class ActivityCreateRoute : AppCompatActivity(),
        OnMapReadyCallback, BottomSheetCreateRoutes.IBottomSheetListener {

    private lateinit var mMap: GoogleMap
    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var mGeocoder: Geocoder
    private lateinit var mTxtStart: TextView
    private lateinit var mTxtEnd: TextView
    private lateinit var mBtnSave: Button

    private var mStartMarker: Marker? = null
    private var mEndMarker: Marker? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_route)
        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        mGeocoder = Geocoder(this)
        mTxtStart = findViewById(R.id.txt_start_address)
        mTxtEnd = findViewById(R.id.txt_end_address)
        mBtnSave = findViewById(R.id.btn_save_route)
        mBtnSave.setOnClickListener { promptForNameAndReturn() }
        mTxtStart.setOnClickListener { loadPlacePicker(REQUEST_PICK_START) }
        mTxtEnd.setOnClickListener { loadPlacePicker(REQUEST_PICK_END) }
        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        mMap.uiSettings.isZoomControlsEnabled = true
        mMap.setOnMapClickListener { latLng -> this@ActivityCreateRoute.handleMapClick(mMap, latLng) }
        mMap.setOnMarkerDragListener(object : GoogleMap.OnMarkerDragListener {
            override fun onMarkerDragStart(marker: Marker) {}
            override fun onMarkerDrag(marker: Marker) {}
            override fun onMarkerDragEnd(marker: Marker) {
                var address: String? = null
                try {
                    val addresses = mGeocoder.getFromLocation(marker.position.latitude, marker.position.longitude, 1)
                    address = addresses[0].getAddressLine(0)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                if (address != null) {
                    marker.position = marker.position
                    marker.snippet = address
                    when (marker) {
                        mEndMarker -> {
                            mEndMarker!!.position = LatLng(marker.position.latitude, marker.position.longitude)
                            mTxtEnd.text = address
                        }
                        mStartMarker -> {
                            mStartMarker!!.position = LatLng(marker.position.latitude, marker.position.longitude)
                            mTxtStart.text = address
                        }
                    }
                    marker.showInfoWindow()
                }
            }
        })
        updateLocation()
    }

    private fun handleMapClick(gMap: GoogleMap?, latLng: LatLng) {
        try {
            val addresses = mGeocoder.getFromLocation(latLng.latitude, latLng.longitude, 1)
            if (addresses.size == 0) return
            val address = addresses[0].getAddressLine(0)
            when {
                mStartMarker == null -> {
                    mStartMarker = gMap!!.addMarker(MarkerOptions()
                            .position(latLng)
                            .title(getString(R.string.start_address_title))
                            .snippet(address)
                            .flat(true)
                            .draggable(true))
                    mStartMarker!!.showInfoWindow()
                    mTxtStart.text = address
                    mTxtStart.visibility = View.VISIBLE
                }
                mEndMarker == null -> {
                    mEndMarker = gMap!!.addMarker(MarkerOptions()
                            .position(latLng)
                            .title(getString(R.string.end_address_title))
                            .snippet(address)
                            .flat(true)
                            .draggable(true))
                    mEndMarker!!.showInfoWindow()
                    mTxtEnd.text = address
                    mTxtEnd.visibility = View.VISIBLE
                    mBtnSave.visibility = View.VISIBLE
                }
                else -> {
                    mStartMarker = null
                    mEndMarker = null
                    gMap!!.clear()
                    mBtnSave.visibility = View.GONE
                    mTxtStart.visibility = View.GONE
                    mTxtEnd.visibility = View.GONE
                }
            }
        } catch (ex: IOException) {
            ex.printStackTrace()
        }


    }

    private fun updateLocation() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_PERM_LOCATION)
        } else {
            fusedLocationClient.lastLocation.addOnSuccessListener(this) { location ->
                if (location != null) {
                    val currentLatLng = LatLng(location.latitude, location.longitude)
                    mMap.isMyLocationEnabled = true
                    mMap.animateCamera(CameraUpdateFactory.newLatLngZoom(currentLatLng, 12f))
                }
            }
        }
    }

    private fun loadPlacePicker(code: Int) {
        val picker = PlacePicker.IntentBuilder()
        try {
            startActivityForResult(picker.build(this@ActivityCreateRoute), code)
        } catch (e: GooglePlayServicesRepairableException) {
            e.printStackTrace()
        } catch (e: GooglePlayServicesNotAvailableException) {
            e.printStackTrace()
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_PICK_START || requestCode == REQUEST_PICK_END) {
            if (resultCode == RESULT_OK) {
                val place = PlacePicker.getPlace(this, data)
                val address = //place.name.toString() + "\n" +
                        place.address.toString()
                val marker: Marker
                val txtView: TextView
                when (requestCode) {
                    REQUEST_PICK_START -> {
                        marker = mStartMarker!!
                        txtView = mTxtStart
                    }
                    else -> {
                        marker = mEndMarker!!
                        txtView = mTxtEnd
                    }
                }
                marker.position = place.latLng
                marker.snippet = address
                txtView.text = address
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            REQUEST_PERM_LOCATION -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    updateLocation()
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun promptForNameAndReturn() {
        val sheet = BottomSheetCreateRoutes().newInstance()
        sheet.show(supportFragmentManager, "Bottom Sheet Dialog Fragment")
    }

    override fun onRouteNameSaved(text: String) {
        finish(text)
    }

    private fun finish(routeName: String) {
        if (mStartMarker != null && mEndMarker != null) {
            var startLocName = "Start"
            var endLocName = "End"
            try {
                startLocName = mGeocoder.getFromLocation(
                        mStartMarker!!.position!!.latitude,
                        mStartMarker!!.position.longitude,
                        1)[0].getAddressLine(0)
                endLocName = mGeocoder.getFromLocation(
                        mEndMarker!!.position.latitude,
                        mEndMarker!!.position.longitude,
                        1)[0].getAddressLine(0)
            } catch (ex: IOException) {
                ex.printStackTrace()
            }

            val returnIntent = Intent()
            val bundle = Bundle()
            bundle.putDoubleArray(START_LOC, doubleArrayOf(mStartMarker!!.position.latitude, mStartMarker!!.position.longitude))
            bundle.putDoubleArray(END_LOC, doubleArrayOf(mEndMarker!!.position.latitude, mEndMarker!!.position.longitude))
            bundle.putString(START_LOC_NAME, startLocName)
            bundle.putString(END_LOC_NAME, endLocName)
            bundle.putString(ROUTE_NAME, routeName)
            returnIntent.putExtra(RESULT_BUNDLE, bundle)
            this.setResult(RESULT_OK, returnIntent)
            finish()
        } else {
            this.setResult(RESULT_CANCELED)
            this.finish()
        }
    }

    companion object {
        private const val REQUEST_PERM_LOCATION = 100
        private const val REQUEST_PICK_START = 200
        private const val REQUEST_PICK_END = 300
        internal const val ROUTE_NAME = "ROUTE_NAME"
        internal const val START_LOC = "START_LOC"
        internal const val END_LOC = "END_LOC"
        internal const val START_LOC_NAME = "START_LOC_NAME"
        internal const val END_LOC_NAME = "END_LOC_NAME"
        internal const val RESULT_BUNDLE = "RESULT_BUNDLE"
    }

}
